File: ios/App/App/Plugins/SmartReplyPlugin.swift
import Foundation
import Capacitor
import MLKitSmartReply

@objc(SmartReplyPlugin)
public class SmartReplyPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier  = "SmartReplyPlugin"
    public let jsName      = "SmartReply"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "suggest", returnType: CAPPluginReturnPromise)
    ]

    private let smartReply = SmartReply.smartReply()

    @objc func suggest(_ call: CAPPluginCall) {
        guard let messagesArray = call.getArray("messages") else {
            call.reject("messages array is required")
            return
        }

        var conversation: [TextMessage] = []

        for item in messagesArray {
            guard
                let msg       = item as? [String: Any],
                let text      = msg["text"]      as? String,
                let isLocal   = msg["isLocal"]   as? Bool,
                let timestamp = msg["timestamp"] as? Double
            else { continue }

            let userId = msg["userId"] as? String ?? "remote_user"

            let message = TextMessage(
                text:        text,
                timestamp:   timestamp / 1000,   // JS uses ms, ML Kit uses seconds
                userID:      userId,
                isLocalUser: isLocal
            )
            conversation.append(message)
        }

        smartReply.suggestReplies(for: conversation) { [weak self] result, error in
            if let error = error {
                call.reject("Smart reply error: \(error.localizedDescription)")
                return
            }

            guard let result = result else {
                call.resolve(["suggestions": [], "status": "no_result"])
                return
            }

            switch result.status {
            case .success:
                let suggestions = result.suggestions.map { $0.text }
                call.resolve(["suggestions": suggestions, "status": "success"])

            case .notSupportedLanguage:
                call.resolve(["suggestions": [], "status": "unsupported_language"])

            default:
                call.resolve(["suggestions": [], "status": "no_reply"])
            }
        }
    }
}

File: android/app/src/main/java/com/edora/app/plugins/SmartReplyPlugin.kt
package com.edora.app.plugins

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.google.mlkit.nl.smartreply.SmartReply
import com.google.mlkit.nl.smartreply.SmartReplySuggestionResult
import com.google.mlkit.nl.smartreply.TextMessage

@CapacitorPlugin(name = "SmartReply")
class SmartReplyPlugin : Plugin() {

    private val smartReplyGenerator = SmartReply.getClient()

    /**
     * Called from JS with:
     * { messages: [{ text: string, isLocal: boolean, userId: string, timestamp: number }] }
     */
    @PluginMethod
    fun suggest(call: PluginCall) {
        val messagesArray = call.getArray("messages") ?: run {
            call.reject("messages array is required")
            return
        }

        val conversation = mutableListOf<TextMessage>()

        try {
            for (i in 0 until messagesArray.length()) {
                val msg = messagesArray.getJSONObject(i)
                val text      = msg.getString("text")
                val isLocal   = msg.getBoolean("isLocal")
                val userId    = msg.optString("userId", "remote_user")
                val timestamp = msg.optLong("timestamp", System.currentTimeMillis())

                val textMessage = if (isLocal) {
                    TextMessage.createForLocalUser(text, timestamp)
                } else {
                    TextMessage.createForRemoteUser(text, timestamp, userId)
                }
                conversation.add(textMessage)
            }
        } catch (e: Exception) {
            call.reject("Failed to parse messages: ${e.message}")
            return
        }

        smartReplyGenerator.suggestReplies(conversation)
            .addOnSuccessListener { result ->
                val suggestions = JSArray()

                when (result.status) {
                    SmartReplySuggestionResult.STATUS_SUCCESS -> {
                        result.suggestions.forEach { suggestion ->
                            suggestions.put(suggestion.text)
                        }
                        val ret = JSObject()
                        ret.put("suggestions", suggestions)
                        ret.put("status", "success")
                        call.resolve(ret)
                    }
                    SmartReplySuggestionResult.STATUS_NOT_SUPPORTED_LANGUAGE -> {
                        val ret = JSObject()
                        ret.put("suggestions", JSArray())
                        ret.put("status", "unsupported_language")
                        call.resolve(ret)
                    }
                    else -> {
                        val ret = JSObject()
                        ret.put("suggestions", JSArray())
                        ret.put("status", "no_reply")
                        call.resolve(ret)
                    }
                }
            }
            .addOnFailureListener { e ->
                call.reject("Smart reply failed: ${e.message}")
            }
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        smartReplyGenerator.close()
    }
}
